test/state/precompiles_cache.cpp

```diff
@@ -0,0 +1,113 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifdef _MSC_VER
+// Disable warning C4996: 'getenv': This function or variable may be unsafe.
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+
+#include "precompiles_cache.hpp"
+#include <nlohmann/json.hpp>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace evmone::state
+{
+namespace
+{
+inline uint64_t fnv1a(bytes_view v) noexcept
+{
+    uint64_t h = 0xcbf29ce484222325;
+    for (const auto b : v)
+        h = (h ^ b) * 0x100000001b3;
+    return h;
+}
+}  // namespace
+
+std::optional<evmc::Result> Cache::find(uint8_t id, bytes_view input, int64_t gas_left) const
+{
+    if (const auto& cache = m_cache.at(id); !cache.empty())
+    {
+        const auto input_hash = fnv1a(input);
+        if (const auto it = cache.find(input_hash); it != cache.end())
+        {
+            if (const auto& o = it->second; !o.has_value())
+                return evmc::Result{EVMC_PRECOMPILE_FAILURE};
+            else
+                return evmc::Result{EVMC_SUCCESS, gas_left, 0, o->data(), o->size()};
+        }
+    }
+    return {};
+}
+
+void Cache::insert(uint8_t id, bytes_view input, const evmc::Result& result)
+{
+    if (id == 4)  // Do not cache "identity".
```

Comment:
It would be nice to use `PrecompileId::identity` here but I assume that you want to avoid including `precompiles.hpp`

Reply:
Done.